kazoo/protocol/connection.py

```diff
@@ -686,68 +674,112 @@ def _connect(self, host, port):
 
-        if client.use_sasl and self.sasl_cli is None:
-            if PURESASL_AVAILABLE:
-                for scheme, auth in client_auth_data_copy:
-                    if scheme == 'sasl':
-                        username, password = auth.split(":")
-                        self.sasl_cli = SASLClient(
-                            host=client.sasl_server_principal,
-                            service='zookeeper',
-                            mechanism='DIGEST-MD5',
-                            username=username,
-                            password=password
-                        )
-                        break
-
-                # As described in rfc
-                # https://tools.ietf.org/html/rfc2831#section-2.1
-                # sending empty challenge
-                self._send_sasl_request(challenge=b'',
-                                        timeout=connect_timeout)
-            else:
-                self.logger.warn('Pure-sasl library is missing while sasl'
-                                 ' authentification is configured. Please'
-                                 ' install pure-sasl library to connect '
-                                 'using sasl. Now falling back '
-                                 'connecting WITHOUT any '
-                                 'authentification.')
-                client.use_sasl = False
-                self._set_connected_ro_or_rw(client)
-        else:
-            self._set_connected_ro_or_rw(client)
-            for scheme, auth in client_auth_data_copy:
-                if scheme == "digest":
-                    ap = Auth(0, scheme, auth)
-                    zxid = self._invoke(
-                        connect_timeout / 1000.0,
-                        ap,
-                        xid=AUTH_XID
-                    )
-                    if zxid:
-                        client.last_zxid = zxid
+        for scheme, auth in client_auth_data_copy:
+            ap = Auth(0, scheme, auth)
+            zxid = self._invoke(connect_timeout / 1000.0, ap, xid=AUTH_XID)
+            if zxid:
+                client.last_zxid = zxid
 
         return read_timeout, connect_timeout
 
-    def _send_sasl_request(self, challenge, timeout):
-        """ Called when sending a SASL request, xid needs be to incremented """
-        sasl_request = SASL(challenge)
-        self._xid = (self._xid % 2147483647) + 1
-        xid = self._xid
-        self._submit(sasl_request, timeout / 1000.0, xid)
-
-    def _set_connected_ro_or_rw(self, client):
-        """ Called to decide whether to set the KeeperState to CONNECTED_RO
-            or CONNECTED"""
-        if self._ro:
-            client._session_callback(KeeperState.CONNECTED_RO)
-            self._ro_mode = iter(self._server_pinger())
-        else:
-            client._session_callback(KeeperState.CONNECTED)
-            self._ro_mode = None
+    def _authenticate_with_sasl(self, host, timeout):
```

Comment:
This was a possibility I thought of when developping the initial DIGEST-MD5 SASL connection but I am not sure if, while authenticating using SASL, the zk server can not send the client any other kind of messages... I also noticed that, in the official java zk client, the choice was made to handle SASL packets in the `readResponses()` loop (see https://github.com/apache/zookeeper/blob/master/zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxn.java#L876)

On your side, what do you think would be the best?

Reply:
I have studied the Java client code and it seems to me that no other packet is allowed while SASL negotiation is in progress. Their code goes to great length to bypass all send queue during that process and that the packets are strict XID for XID request/response exchanges: 
- https://github.com/apache/zookeeper/blob/caca062767c36525e6ecead2ae0f34c447394809/zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxnSocketNIO.java#L170
- https://github.com/apache/zookeeper/blob/master/zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxn.java#L890
- https://github.com/apache/zookeeper/blob/cb9f303bda9137d1aebe8eff3eab85c8a59f3cdd/zookeeper-server/src/main/java/org/apache/zookeeper/client/ZooKeeperSaslClient.java#L296
 

Furthermore, the client also does check that the state is `CONNECTED` or `CONNECTEDREADONLY` *before* starting the SASL authentication:

- https://github.com/apache/zookeeper/blob/master/zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxn.java#L1163
- https://github.com/apache/zookeeper/blob/22e4acc802904231d537ae18f479cfc729851d87/zookeeper-server/src/main/java/org/apache/zookeeper/ZooKeeper.java#L638

This indicate the the SASL negotiation only starts once the connection to the server is established (which means a `CONNECTED` event was already emitted), which my PR emulates. Note that native forms of authentication, when passed in as `auth_data`, are also evaluated right after the client reaches the connected state.
The only possible downside to my approach is a connection watcher would get a `CONNECTED` event followed by a `LOST` which is already possible (networks can fail) and what happens with native authentication schemes.
Looking at the links above, what do you think? Do you agree with my interpretation?

As for factoring out the SASL negotiation in its own function, since only SASL packets can be sent during the authentication process anyway, and only during connection, I do not see how this could bite us down the road  (and it definitely made my PR less invasive).
If you still feel strongly about this, can you give me guidance as to how you would tackle integrating the SASL loop, with the above state conditions, in the main send/receive loop?
Also, could this be done in a different refactoring PR?

Reply:
Thanks for all those links, you are right, I missed the `findSendablePacket` method and it really seems nothing can happen during the SASL authentication. It makes totally sense to do the things the way you suggest.